dot/state/storage_notify.go

```diff
@@ -33,55 +34,106 @@ type SubscriptionResult struct {
 	Changes []KeyValue
 }
 
-//StorageSubscription holds data for Subscription to Storage
-type StorageSubscription struct {
-	Filter   map[string]bool
-	Listener chan<- *SubscriptionResult
+// Observer interface defines functions needed for observers, Observer Design Pattern
+type Observer interface {
+	Update(result *SubscriptionResult)
+	GetID() uint
+	GetFilter() map[string][]byte
 }
 
-// RegisterStorageChangeChannel function to register storage change channels
-func (s *StorageState) RegisterStorageChangeChannel(sub StorageSubscription) (byte, error) {
-	s.changedLock.RLock()
+// RegisterStorageObserver to add abserver to notification list
+func (s *StorageState) RegisterStorageObserver(o Observer) {
+	s.observerList = append(s.observerList, o)
 
-	if len(s.subscriptions) == 256 {
-		return 0, errors.New("storage subscriptions limit reached")
+	// notifyObserver here to send storage value of current state
+	sr, err := s.blockState.BestBlockStateRoot()
+	if err != nil {
+		logger.Debug("error registering storage change channel", "error", err)
 	}
-
-	var id byte
-	for {
-		id = generateID()
-		if s.subscriptions[id] == nil {
-			break
+	go func() {
+		if err := s.notifyObserver(sr, o); err != nil {
```

Comment:
will not returning cause an error since `sr` will be undefined?

Reply:
good point, I added a return.